// eslint-plugin-clean-regex/dist/rules/no-octal-escape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rules_util_1 = require("../rules-util");
const ast_util_1 = require("../ast-util");
exports.default = {
    meta: {
        type: "problem",
        docs: {
            description: "Disallow octal escapes outside of character classes.",
            url: rules_util_1.getDocUrl("no-octal-escape"),
        },
    },
    create(context) {
        return rules_util_1.createRuleListener(({ visitAST, reportElement }) => {
            visitAST({
                onCharacterEnter(node) {
                    if (node.parent.type === "CharacterClass" || node.parent.type === "CharacterClassRange") {
                        return;
                    }
                    if (node.value > 0 && ast_util_1.isOctalEscapeSequence(node)) {
                        context.report(Object.assign({ message: "Do not use octal escapes because they might be confused with backreferences." }, reportElement(node)));
                    }
                },
            });
        });
    },
};
