// eslint-plugin-clean-regex/dist/rules/no-potentially-empty-backreference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rules_util_1 = require("../rules-util");
const ast_util_1 = require("../ast-util");
exports.default = {
    meta: {
        type: "problem",
        docs: {
            description: "Disallow backreferences that reference a group that might not be matched.",
            url: rules_util_1.getDocUrl("no-potentially-empty-backreference"),
        },
    },
    create(context) {
        return rules_util_1.createRuleListener(({ visitAST, reportElement }) => {
            visitAST({
                onBackreferenceEnter(node) {
                    if (ast_util_1.isEmptyBackreference(node)) {
                        return;
                    }
                    if (!ast_util_1.backreferenceAlwaysAfterGroup(node)) {
                        context.report(Object.assign({ message: "Some path leading to the backreference do not go through the referenced capturing group without resetting its text." }, reportElement(node)));
                    }
                },
            });
        });
    },
};
